### PR TITLE
Add XML support to clientcore

### LINKED ISSUE #42718: XML support
We need to ensure that we have xml support in place, for libraries such as storage.

### PR DESCRIPTION
**Pull request by alzimmermsft**
# Description

Resolves #42718 

Adds XML support to `io.clientcore:core`.

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/clientcore/core/pom.xml [hunk 1]
```diff
@@ -181,6 +183,32 @@
...
+        <version>1.0.0-beta.10</version> <!-- {x-version-update;com.azure.tools:codesnippet-maven-plugin;external_dependency} -->
+        <configuration>
+          <additionalCodesnippets>
+            <additionalCodesnippet>
+              <root>src/test</root>
+              <glob>**/*.java</glob>
+            </additionalCodesnippet>
+          </additionalCodesnippets>
+        </configuration>
+      </plugin>
```
Comment (srnagar) [inline sdk/clientcore/core/pom.xml:199]: Should this plugin be in the parent pom?

File: sdk/clientcore/core/src/main/java/io/clientcore/core/serialization/xml/XmlSerializable.java [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+     * @param xmlReader The {@link XmlReader} being read.
+     * @param rootElementName Optional root element name to override the default defined by the model. Used to support
+     * cases where the model can deserialize from different root element names.
+     * @param <T> The type of the object.
+     * @return The object that the XML stream represented, may return null.
+     * @throws XMLStreamException If an object fails to be read from the {@code xmlReader}.
+     */
+    static <T extends XmlSerializable<T>> T fromXml(XmlReader xmlReader, String rootElementName)
+        throws XMLStreamException {
+        throw new UnsupportedOperationException("Implementation of XmlSerializable must define this factory method.");
```
Comment (srnagar) [inline sdk/clientcore/core/src/main/java/io/clientcore/core/serialization/xml/XmlSerializable.java:184]: Why do we need these static methods in the interface? We can't have any default implementation, so, is it required to define these methods here?

File: sdk/clientcore/core/src/main/java/io/clientcore/core/serialization/xml/XmlReadValueCallback.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+import javax.xml.stream.XMLStreamException;
+
+/**
+ * A callback used when reading an XML value, such as {@link XmlReader#getNullableElement(XmlReadValueCallback)}.
+ *
+ * @param <T> Input type of the callback.
+ * @param <R> Output type of the callback.
+ */
+@FunctionalInterface
+public interface XmlReadValueCallback<T, R> {
```
Comment (srnagar) [inline sdk/clientcore/core/src/main/java/io/clientcore/core/serialization/xml/XmlReadValueCallback.java:15]: Should we use `Function` here to match the JDK types (similar to `Consumer` above)? 
`XmlReadValueFunction`.

**Review by srnagar — APPROVED** [review]

File: sdk/clientcore/core/pom.xml (hunk 1 above)
Comment (alzimmermsft) [reply sdk/clientcore/core/pom.xml:199]: The plugin is in the parent POM, this is extending the configurations to pull snippets from test code. Or, are you saying we should put the configuration to add test code as snippet sources to the parent POM?

File: sdk/clientcore/core/src/main/java/io/clientcore/core/serialization/xml/XmlSerializable.java (hunk 2 above)
Comment (alzimmermsft) [reply sdk/clientcore/core/src/main/java/io/clientcore/core/serialization/xml/XmlSerializable.java:184]: This was a basic copy from azure-xml. It's similar to the static methods in `JsonSerializable` where there is an expectation that this needs to be overwritten.

File: sdk/clientcore/core/src/main/java/io/clientcore/core/serialization/xml/XmlReadValueCallback.java (hunk 3 above)
Comment (alzimmermsft) [reply sdk/clientcore/core/src/main/java/io/clientcore/core/serialization/xml/XmlReadValueCallback.java:15]: Do you mean `implement Function`? If so, that can't be done as this adds a throws for `XMLStreamException` to the functional method, allowing for all the try-catches to runtime exceptions to be removed in lambda code.

Comment (srnagar) [reply sdk/clientcore/core/src/main/java/io/clientcore/core/serialization/xml/XmlReadValueCallback.java:15]: No, I was just suggesting updating the name to use the suffix `Function` to indicate that there is an input and an output similar to [Function](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/function/Function.html).

Updated name - `XmlReadValueFunction`

File: sdk/clientcore/core/pom.xml (hunk 1 above)
Comment (srnagar) [reply sdk/clientcore/core/pom.xml:199]: Yeah, we should make this the default to allow test code to be snippet sources.

File: sdk/clientcore/core/src/main/java/io/clientcore/core/serialization/xml/XmlSerializable.java (hunk 2 above)
Comment (srnagar) [reply sdk/clientcore/core/src/main/java/io/clientcore/core/serialization/xml/XmlSerializable.java:184]: I don't think this is enforceable for implementation types. So, not sure how adding these static methods will help. Should we also remove the static method in `JsonSerializable` (just in clientcore)?

**Pull request merged**